### ISSUE TITLE
"Probe returned with Transport endpoint is not connected" when creating a new topology

### ISSUE THREAD

**Issue by bsod90** [heketi/heketi#945]
Hi Heketi project,

I'm trying to deploy heketi & glusterfs on my k8s cluster and I'm running into a rather weird issue:

I'm trying to use 4 nodes for my gluster (192.168.2.110, 111, 112 and 113)
I'm following this tutorial: https://github.com/heketi/heketi/blob/master/doc/admin/install-kubernetes.md

And at the step, where it finally tries to create a topology, I'm getting this output:
```
[admin@master kubernetes]$ heketi-cli topology load --json=topology-ts.json
        Found node 110.k8s on cluster 3587a322ad49ecee9aebb9a71c1a4369
                Found device /dev/sdc
        Creating node 111.k8s ... Unable to create node: Unable to execute command on glusterfs-jf5bd: peer probe: failed: Probe returned with Transport endpoint is not connected
        Creating node 112.k8s ... Unable to create node: Unable to execute command on glusterfs-jf5bd: peer probe: failed: Probe returned with Transport endpoint is not connected
        Creating node 113.k8s ... Unable to create node: Unable to execute command on glusterfs-jf5bd: peer probe: failed: Probe returned with Transport endpoint is not connected
```

I tried digging a little bit into this issue, at first I thought it's some networking stuff, like a Pod can't talk to k8s nodes or smth, but that doesn't seem to be a case. Iptable are also pretty permissive (it's Centos 7 with firewalld disabled and the rules from the tutorial added, just in case).

But then I've noticed this weird behavior:

```
[admin@master ~]$ kubectl exec -it glusterfs-jf5bd /bin/bash
[root@110 /]# gluster peer probe 192.168.2.112
peer probe: success. Host 192.168.2.112 port 24007 already in peer list
[root@110 /]# exit
You have mail in /var/spool/mail/admin
[admin@master ~]$ kubectl exec -it glusterfs-jf5bd gluster peer probe 192.186.2.112
peer probe: failed: Probe returned with Transport endpoint is not connected
[admin@master ~]$
```

Basically, if I `kubectl exec` into the gluster pod on .110 and run `gluster peer probe 192.168.2.111, 112, 113` it succeeds
But if I do that directly by passing `gluster peer probe 192.168.2.*` as an argument to `kubectl exec`, just like it's done in heketi code, it fails with the same error message as in the above log.

This made me confused, as I can't come up with any reasonable explanation on what may be causing this behavior. The only thing I may think of is some timeouts voodoo.

I'm using kubernetes 1.7.3 and all the heketi scripts from master (heketi-cli I've copied from the bootstrap pod).

I would appreciate any suggestions on how to debug this or, maybe, some manual hacks to get past the glusterfs deployment.

Thank you!

**Comment by bsod90** [comment]
After the second day of fighting I found a weird workaround: if I replace storage IPs with hostnames, it magically starts to work.

This allowed me to proceed almost to the finish line, but I ran into another inconsistency almost at the end: at the PVC fulfilment stage.

```[admin@master kubernetes]$ kubectl describe pvc export-maxim-stateful-0
Name:           export-maxim-stateful-0
Namespace:      default
StorageClass:   slow
Status:         Pending
Volume:
Labels:         app=maxim-stateful
Annotations:    volume.beta.kubernetes.io/storage-class=slow
                volume.beta.kubernetes.io/storage-provisioner=kubernetes.io/glusterfs
Capacity:
Access Modes:
Events:
  FirstSeen     LastSeen        Count   From                            SubObjectPath   Type            Reason                  Message
  ---------     --------        -----   ----                            -------------   --------        ------                  -------
  32m           32m             1       persistentvolume-controller                     Warning         ProvisioningFailed      storageclass.storage.k8s.io "slow" not found
  29m           27m             3       persistentvolume-controller                     Warning         ProvisioningFailed      Failed to provision volume with StorageClass "slow": glusterfs: create volume err: error creating volume Unable to execute command on glusterfs-lwxgf:   Volume group "vg_ca714dd674b6845ded3687725b61a78f" not found
  Cannot process volume group vg_ca714dd674b6845ded3687725b61a78f
  29m   27m     6       persistentvolume-controller             Warning ProvisioningFailed      Failed to provision volume with StorageClass "slow": glusterfs: create volume err: error creating volume Unable to execute command on glusterfs-5mcdl:   Volume group "vg_519b6710479803d08801dfd1ed2cdc1a" not found
  Cannot process volume group vg_519b6710479803d08801dfd1ed2cdc1a
  32m   24m     13      persistentvolume-controller             Warning ProvisioningFailed      Failed to provision volume with StorageClass "slow": glusterfs: create volume err: error creating volume Unable to execute command on glusterfs-0rgsk:   Volume group "vg_f8df7e130ec5176810dd31e79c51dfeb" not found
  Cannot process volume group vg_f8df7e130ec5176810dd31e79c51dfeb
  31m   24m     7       persistentvolume-controller             Warning ProvisioningFailed      Failed to provision volume with StorageClass "slow": glusterfs: create volume err: error creating volume Unable to execute command on glusterfs-hkw6h:   Volume group "vg_e183ae78fed4263bf224f1ce73e0f2e0" not found
  Cannot process volume group vg_e183ae78fed4263bf224f1ce73e0f2e0
  24m   24m     1       persistentvolume-controller             Warning ProvisioningFailed      Failed to provision volume with StorageClass "slow": glusterfs: create volume err: error creating volume Get http://10.108.110.115:8080/queue/9d29b59acba159f4c3ad5f1c365da6ea: dial tcp 10.108.110.115:8080: getsockopt: connection refused
  19m   18m     2       persistentvolume-controller             Warning ProvisioningFailed      Failed to provision volume with StorageClass "slow": glusterfs: create volume err: error creating volume Post http://10.108.110.115:8080/volumes: dial tcp 10.108.110.115:8080: i/o timeout
  34m   17m     13      persistentvolume-controller             Warning ProvisioningFailed      Failed to provision volume with StorageClass "slow": glusterfs: create volume err: error creating volume Post http://heketi:8080/volumes: dial tcp: lookup heketi on 192.168.2.200:53: no such host
  23m   28s     42      persistentvolume-controller             Warning ProvisioningFailed      Failed to provision volume with StorageClass "slow": glusterfs: create volume err: failed to create endpoint/service error creating endpoint: Endpoints "glusterfs-dynamic-export-maxim-stateful-0" is invalid: [subsets[0].addresses[0].ip: Invalid value: "112.k8s": must be a valid IP address, (e.g. 10.9.8.7), subsets[0].addresses[1].ip: Invalid value: "111.k8s": must be a valid IP address, (e.g. 10.9.8.7), subsets[0].addresses[2].ip: Invalid value: "110.k8s": must be a valid IP address, (e.g. 10.9.8.7), subsets[0].addresses[3].ip: Invalid value: "113.k8s": must be a valid IP address, (e.g. 10.9.8.7)]
You have mail in /var/spool/mail/admin
[admin@master kubernetes]$```

I tried to find where it gets those hostnames (110-113.k8s) and found this:
https://github.com/kubernetes/kubernetes/blame/master/pkg/volume/glusterfs/glusterfs.go#L898

Documentation here https://github.com/heketi/heketi/blob/master/doc/api/api.md#add-node suggests that `storage` array may contain hostnames and even recommends putting hostnames there, however, storage controller from k8s seems to be expecting explicit IPs there.
But if I use IPs in my node definition, then the initial configuration doesn't work, due to the weird issue mentioned above :|

**Comment by zhaorendong** [comment]
Can I have a look at your topology-ts.json? Make sure that  manage use the name of k8s node and storage use the node IP
